src/ngraph/pass/opset1_transform.cpp

```diff
@@ -0,0 +1,101 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include "ngraph/pass/opset1_transform.hpp"
+#include "ngraph/graph_util.hpp"
+#include "ngraph/op/get_output_element.hpp"
+#include "ngraph/op/softmax.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+// @TODO: Shouldn't this be moved to a common utility class? This mapping to OP_TYPEID gets repeated many times.
+#define NGRAPH_OP(a, b) a,
+enum class OP_TYPEID
+{
+#include "ngraph/op/op_tbl.hpp"
+};
+#undef NGRAPH_OP
+
+#define NGRAPH_OP(a, b) {#a, OP_TYPEID::a},
+static unordered_map<string, OP_TYPEID> typeid_map{
+#include "ngraph/op/op_tbl.hpp"
+};
+#undef NGRAPH_OP
+
+static OP_TYPEID get_typeid(shared_ptr<Node> node)
+{
+    OP_TYPEID type_id;
+    auto it = typeid_map.find(node->description());
+    if (it != typeid_map.end())
+    {
+        type_id = it->second;
+    }
+    else
+    {
+        throw unsupported_op("Unsupported op '" + node->description() + "'");
+    }
+    return type_id;
+}
+// END mapping to OP_TYPEID
+
+bool pass::Opset1Transformation::run_on_node(shared_ptr<Node> node)
+{
+    bool modified = false;
+
+    size_t opset_version = node->get_opset_version();
+
+    if (opset_version == 1)
+    {
+        return modified;
+    }
+
+    if (opset_version != 0)
+    {
+        throw ngraph_error("Opset 1 transformation pass failed for " + node->get_name() +
+                           ", only opset 0 operations expected. Opset " + to_string(opset_version) +
+                           " found.");
+    }
+
+    switch (get_typeid(node))
+    {
+    case OP_TYPEID::Softmax:
+    {
+        auto tmp = dynamic_cast<const op::set0::Softmax*>(node.get());
+        AxisSet axes = tmp->get_axes();
+
+        if (axes.size() != 1)
+        {
+            throw ngraph_error(
+                "Unable to convert Softmax:0 to Softmax:1 with more then one axis. " +
+                node->get_name() + ".");
+        }
+
+        auto replacement_node =
+            make_shared<op::set1::Softmax>(node->input(0).get_source_output(), axes.to_vector()[0]);
+        replace_node(node, replacement_node);
+        modified = true;
+        break;
+    }
+    default:
+    {
+        node->set_opset_version(1);
```

Comment:
I also thought about using the `friend` keyword and hiding the setter.

Reply:
I am not sure about using `friend` mechanism in this case.
In order to have access to `set_opset_version` from level of `Opset1Transformation::run_on_node` it will be necessary to declare `Opset1Transformation::run_on_node` as `friend` in `Node` class. I think that adding `friend` for all opset transformation will be not good solution. 
What's more declaration `run_on_node' as `friend` in `Node` requires that `Opset1Transformation::run_on_node` is declared earlier. I am not sure how to ensure it (class forward declaration is not enough in this case).
One solution can be declaration **class** `Opset1Transformation` as `friend` in `Node`. Cons: access to all private members of `Node` and necessity to declare `friend` to all opset transformations (we can not declare `NodePass` as `friend class` because it does not work for inheritance).
Second: We can implement **function** with `friend` privileges to `Node` which will be changing opset version in other file which will be included only in transform `.cpp` files.
What do you think about it?

As I know declaring  **virtual** 'run_on_node' from

Reply:
I suppose we could do something like this...

```
template <int VERSION>
friend class OpsetTransformation;
```

In any case, I think if we deal with the more fundamental issue (that a single opset version field attached to a node may not be sufficient), we may be able to avoid the issue altogether. I was chatting with @diyessi about this at our usual 1:1 today, and I think he has some good ideas here so let's ask him to chime in.